### PR TITLE
Android: Reimplement input binding

### PR DESCRIPTION
**Pull request by SeannyM**
I fixed up the old stuff to make it work with the new settings UI. Functionally, everything is as it was before, plus a new clear button to remove bindings. I changed the strings to match the desktop UI, which was also convenient because it lead to less strings. :)

Here's a screenshot just because:
screenshot oct 30 2016 7-54-09 pm
(^ When you map a real Wiimote to an emulated one because you don't have a DolphinBar :P)

---

This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4410)

### CODE REVIEW

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java [hunk 1]
```diff
@@ -291,13 +313,395 @@ private void addHackSettings(ArrayList<SettingsItem> sl)
...
+		sl.add(new HeaderSetting(null, null, R.string.controller_dpad, 0));
+		sl.add(new InputBindingSetting(SettingsFile.KEY_WIIBIND_DPAD_UP + wiimoteNumber, SettingsFile.SECTION_BINDINGS, R.string.generic_up, bindDPadUp));
+		sl.add(new InputBindingSetting(SettingsFile.KEY_WIIBIND_DPAD_DOWN + wiimoteNumber, SettingsFile.SECTION_BINDINGS, R.string.generic_down, bindDPadDown));
+		sl.add(new InputBindingSetting(SettingsFile.KEY_WIIBIND_DPAD_LEFT + wiimoteNumber, SettingsFile.SECTION_BINDINGS, R.string.generic_left, bindDPadLeft));
+		sl.add(new InputBindingSetting(SettingsFile.KEY_WIIBIND_DPAD_RIGHT + wiimoteNumber, SettingsFile.SECTION_BINDINGS, R.string.generic_right, bindDPadRight));
+	}
 
-		sl.add(new CheckBoxSetting(SettingsFile.KEY_GCADAPTER_RUMBLE + gcPadNumber, SettingsFile.SECTION_CORE, R.string.gc_adapter_rumble, R.string.gc_adapter_rumble_description, false, rumble));
-		sl.add(new CheckBoxSetting(SettingsFile.KEY_GCADAPTER_BONGOS + gcPadNumber, SettingsFile.SECTION_CORE, R.string.gc_adapter_bongos, R.string.gc_adapter_bongos_description, false, bongos));
+	private void addExtensionSubSettings(ArrayList<SettingsItem> sl, int wiimoteNumber, int extentionType)
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java]: `extensionType`

**Comment by degasus** [comment]
@sigmabeta Do you want to help me with reviewing?

**Comment by sigmabeta** [comment]
I can take a look later tonight.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java [hunk 2]
```diff
@@ -291,13 +313,395 @@ private void addHackSettings(ArrayList<SettingsItem> sl)
...
+		Setting bindShakeY = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_SHAKE_Y + wiimoteNumber);
+		Setting bindShakeZ = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_SHAKE_Z + wiimoteNumber);
+		Setting bindDPadUp = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_DPAD_UP + wiimoteNumber);
+		Setting bindDPadDown = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_DPAD_DOWN + wiimoteNumber);
+		Setting bindDPadLeft = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_DPAD_LEFT + wiimoteNumber);
+		Setting bindDPadRight = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIBIND_DPAD_RIGHT + wiimoteNumber);
+
+		sl.add(new SingleChoiceSetting(SettingsFile.KEY_WIIMOTE_EXTENSION + (wiimoteNumber - 3), SettingsFile.SECTION_BINDINGS, R.string.wiimote_extensions, R.string.wiimote_extensions_descrip, R.array.wiimoteExtensionsEntries, R.array.wiimoteExtensionsValues, 0, extension));
```
Comment (degasus) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java]: same here?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java [hunk 3]
```diff
@@ -291,13 +313,395 @@ private void addHackSettings(ArrayList<SettingsItem> sl)
...
+			sl.add(new CheckBoxSetting(SettingsFile.KEY_GCADAPTER_RUMBLE + gcPadNumber, SettingsFile.SECTION_CORE, R.string.gc_adapter_rumble, R.string.gc_adapter_rumble_description, false, rumble));
+			sl.add(new CheckBoxSetting(SettingsFile.KEY_GCADAPTER_BONGOS + gcPadNumber, SettingsFile.SECTION_CORE, R.string.gc_adapter_bongos, R.string.gc_adapter_bongos_description, false, bongos));
+		}
+	}
+
+	private void addWiimoteSubSettings(ArrayList<SettingsItem> sl, int wiimoteNumber)
 	{
-		Setting rumble = mSettings.get(SettingsFile.SECTION_CORE).getSetting(SettingsFile.KEY_GCADAPTER_RUMBLE + gcPadNumber);
-		Setting bongos = mSettings.get(SettingsFile.SECTION_CORE).getSetting(SettingsFile.KEY_GCADAPTER_BONGOS + gcPadNumber);
+		Setting extension = mSettings.get(SettingsFile.SECTION_BINDINGS).getSetting(SettingsFile.KEY_WIIMOTE_EXTENSION + (wiimoteNumber - 3));
```
Comment (SeannyM) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/settings/SettingsFragmentPresenter.java]: The Wii binding settings continue off the GameCube settings for controller number. GC controllers are 0-3, Wii controllers are 4-7, so WiimoteA_4 in the main ini is the mapping for Wiimote 1 button A. In the Wii ini, where we set the extension, the controllers are 1-4.

**Comment by SeannyM** [comment]
Rebased and removed the hacky stuff, everything should be working fine now.

**Comment by Brotanium** [comment]
I have discovered a problem.  If you bind the controls the first time everything works, but if you bind them again you have to bind every control again or the controls that you did not rebind do not function.

**Comment by MaJoRoesch** [comment]
@SeannyM We really need more for the progress report! Does it need review? Is this ready for merging?

**Comment by SeannyM** [comment]
@Brotanium good catch, I fixed the issue. Regarding unbinding via controller, it's something I can look into sometime in the future; at some point I'd like to design a better UI for this anyways, because right now using the same UI as the rest of the settings means there is a lot of scrolling.

@MaJoRoesch if nobody else has any issues with this, it's good to go.

**Comment by Helios747** [comment]
*[Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/MotionAlertDialog.java, line 26 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4410#-KXXvBjrMmJU-d1AGosO:-KXXvBjs44PztsY1unpy:b-q4qajs) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4a2d26d5daa371dbacf9b5ac60ec17d0d9b6f9e2/Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/MotionAlertDialog.java#L26)):*
> ```Java
> {
> 	// The selected input preference
> 	private final InputBindingSetting item;
> ```

Can we name this something else? 'item' isn't particularly descriptive of what this is for even if the code it's used in is simple.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4410)*

**Comment by SeannyM** [comment]
*[Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/MotionAlertDialog.java, line 26 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4410#-KXXvBjrMmJU-d1AGosO:-KXXw3MQHgvvjNWNzNMS:b900i2t) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4a2d26d5daa371dbacf9b5ac60ec17d0d9b6f9e2/Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/MotionAlertDialog.java#L26)):*
<details><summary><i>Previously, Helios747 (Anthony) wrote…</i></summary>
> Can we name this something else? 'item' isn't particularly descriptive of what this is for even if the code it's used in is simple.
</details>
I agree "item" isn't a very fitting name for this, I only named it that because they are called "item" in onClickListeners. Though I guess it's more fitting in that situation, because the setting is the item you are clicking on. I'll change it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4410)*

**Comment by Helios747** [comment]
:lgtm:

---

Reviewed 3 of 16 files at r1, 11 of 12 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4410#-:-KXY4Fc6fqAHQG9qpwbg:bnfp4nl)*

**Comment by MaJoRoesch** [comment]
@SeannyM Tested. I ran into a critical issue. The controller presses are still being picked up by the UI even when in game! When I press B on the controller in order to press a button in game, it backs out of the game and back to the UI. I can't use my controller at all!

**Comment by MaJoRoesch** [comment]
Aaaand it's already merged. Ok, well. One last comment. This is probably outside the scope of this PR, but there isn't any way to put in L/R analog. Many games use them as separate buttons (rogue leader, super mario sunshine), or require pressing analog before digital will be accepted (metroid prime). So it's important for it to have mappable analog L/R separate from the digital L/R.

**Comment by SeannyM** [comment]
@MaJoRoesch regarding the UI input issue, I can't reproduce it on my end. Can you check to make sure the bindings are saved? IIRC, someone else had the same issue and it turned out they were using the 'X' to close the settings, which actually exits without saving. To save the settings, you have to back out of the settings fully without using the 'X'. If this isn't the problem, my only guess would be differences between Android versions, maybe focus isn't being given to the emulation activity?

**Comment by MaJoRoesch** [comment]
I went back to check at the time, and it was still bound. And I just rechecked it, and the settings were all still there. And I literally turned on the phone, and ran Dolphin. There is absolutely nothing else going on!

**Comment by SeannyM** [comment]
Perhaps we should prevent the back button from exiting emulation, and instead put an exit button in the dropdown menu? Or at least an option to disable the back button. This would also help prevent accidental exits.

**Comment by sigmabeta** [comment]
Make sure to only do this on touchscreen devices though.

**Comment by Brotanium** [comment]
So I tested again to see if I could get multiplayer working and the unbinding problem persists making multiplayer impossible since it unbinds when setting another controller.  I then tried to rebind a single controller and all the other controls unbinding when done.  Another thing to note a new Panic Alert occurs now which says "Panic Alert: IOS HLE: Couldn't get an absolute path; the root directory will be returned. This will most likely lead to failures.".  I am not sure if this is related to anything done but I was just wondering.  Thanks.

**Comment by JosJuice** [comment]
@Brotanium I broke IOS HLE in 5.0-1339 except on Windows, so many Wii games don't work right now. A fix is in progress.

**Comment by SeannyM** [comment]
@Brotanium I just tested, and I can map multiple controllers at once without a problem. I'm not sure why it wouldn't be working for you. You're getting the toast message saying the settings are saved, right? It should show up after you leave the settings and return to the game list.

**Comment by Brotanium** [comment]
After uninstalling dolphin then reinstalling it GameCube controller work
fine with no unbinding problems but Wii games say the Wii remotes are
disconnected by emulating software so I cannot test if the unbinding
problem persists.

**Pull request merged**